rust-runtime/aws-smithy-http-server/src/routing/tiny_map.rs

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::{
+    borrow::Borrow,
+    collections::{hash_map::RandomState, HashMap},
+    fmt,
+    hash::{BuildHasher, Hash},
+};
+
+/// A map implementation with fast iteration which switches backing storage from [`Vec`] to
+/// [`HashMap`] when the number of entries exceeds `CUTOFF` (which defaults to 20).
+#[derive(Clone)]
+pub struct TinyMap<K, V, S = RandomState, const CUTOFF: usize = 20> {
+    inner: TinyMapInner<K, V, S, CUTOFF>,
+}
+
+impl<K, V, S, const CUTOFF: usize> fmt::Debug for TinyMap<K, V, S, CUTOFF>
+where
+    K: fmt::Debug,
+    V: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("TinyMap").field("inner", &self.inner).finish()
+    }
+}
+
+#[derive(Clone)]
+enum TinyMapInner<K, V, S, const CUTOFF: usize> {
+    Vec(Vec<(K, V)>),
+    HashMap(HashMap<K, V, S>),
+}
+
+impl<K, V, S, const CUTOFF: usize> fmt::Debug for TinyMapInner<K, V, S, CUTOFF>
+where
+    K: fmt::Debug,
+    V: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Vec(arg0) => f.debug_tuple("Vec").field(arg0).finish(),
+            Self::HashMap(arg0) => f.debug_tuple("HashMap").field(arg0).finish(),
+        }
+    }
+}
+
+enum EitherIterator<Left, Right> {
+    Left(Left),
+    Right(Right),
+}
+
+impl<Left, Right> Iterator for EitherIterator<Left, Right>
+where
+    Left: Iterator,
+    Right: Iterator<Item = Left::Item>,
+{
+    type Item = Left::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::Left(left) => left.next(),
+            Self::Right(right) => right.next(),
+        }
+    }
+}
+
+/// An owning iterator over the entries of a `TinyMap`.
+///
+/// This struct is created by the [`into_iter`](IntoIterator::into_iter) method on [`TinyMap`] (
+/// provided by the [`IntoIterator`] trait). See its documentation for more.
+pub struct IntoIter<K, V> {
+    inner: EitherIterator<std::vec::IntoIter<(K, V)>, std::collections::hash_map::IntoIter<K, V>>,
+}
+
+impl<K, V> Iterator for IntoIter<K, V> {
+    type Item = (K, V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+}
+
+impl<K, V, S, const CUTOFF: usize> IntoIterator for TinyMap<K, V, S, CUTOFF> {
+    type Item = (K, V);
+
+    type IntoIter = IntoIter<K, V>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        let inner = match self.inner {
+            TinyMapInner::Vec(vec) => EitherIterator::Left(vec.into_iter()),
+            TinyMapInner::HashMap(hash_map) => EitherIterator::Right(hash_map.into_iter()),
+        };
+        IntoIter { inner }
+    }
+}
+
+impl<K, V, S, const CUTOFF: usize> FromIterator<(K, V)> for TinyMap<K, V, S, CUTOFF>
+where
+    K: Hash + Eq,
+    S: BuildHasher + Default,
+{
+    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> Self {
```

Comment:
This implementation is very simple.

A different implementation which might be a bit faster is that we first create a `[MaybeUninit<(K, V)>; CUTOFF]`, populate it, if it overflows then `into_iter()` it into a `HashMap`, if it doesn't then `.to_vec()` the populated slice. This would avoid the `Vec` allocation in the case where it overflows.

The downside to this approach is that it'd require `unsafe`. We could use `Option` instead of `MaybeUninit` but this might cause performance hit over `MaybeUninit`.

Reply:
Yeah agree.

I was referring to https://github.com/awslabs/smithy-rs/blob/007f08ca73b5f6b9715dfa180ada91f09f68045a/rust-runtime/aws-smithy-http-server/src/routing/mod.rs#L161-L197 (link before doesn't expand out the code, just highlights it in secret). Here we could probably avoid the `collect` with an `iter_mut` for example, but it might not be worth it for the same reason.